### PR TITLE
Fix calculate accuracy problem in log loss metric.

### PR DESCRIPTION
**Pull request by ZhangYuef**
**Finish**
- change data precision
- change eps setting in log loss calculation

**Aim**
- To solve issue #55 .
- To get a robust activation layer function.

**TODO**
- [x] Support all kinds of built-in activation layers in `PyTorch`.

### CODE REVIEW

**Comment by ZhangYuef** [comment]
A script to test the result after refactoring activation layers and adding Dice activation function:

In `core.py `:
```python
if __name__ == "__main__":
    from torchsummary import summary
    torch.manual_seed(7)

    a = DNN(2, [200, 80, 1], activation='dice')
    summary(a, input_size=(2,))
    b = torch.rand((3, 2))
    c = a(b)
    print(c.size())
    print(f"b: {b}")
    print(f"c: {c}")
```
>python core.py

Output:
```
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
================================================================
            Linear-1                  [-1, 200]             600
       BatchNorm1d-2                  [-1, 200]             400
           Sigmoid-3                  [-1, 200]               0
              Dice-4                  [-1, 200]               0
           Dropout-5                  [-1, 200]               0
            Linear-6                   [-1, 80]          16,080
       BatchNorm1d-7                   [-1, 80]             160
           Sigmoid-8                   [-1, 80]               0
              Dice-9                   [-1, 80]               0
          Dropout-10                   [-1, 80]               0
           Linear-11                    [-1, 1]              81
      BatchNorm1d-12                    [-1, 1]               2
          Sigmoid-13                    [-1, 1]               0
             Dice-14                    [-1, 1]               0
          Dropout-15                    [-1, 1]               0
================================================================
Total params: 17,323
Trainable params: 17,323
Non-trainable params: 0
----------------------------------------------------------------
Input size (MB): 0.00
Forward/backward pass size (MB): 0.01
Params size (MB): 0.07
Estimated Total Size (MB): 0.08
----------------------------------------------------------------
torch.Size([3, 1])
b: tensor([[0.9537, 0.8912],
        [0.8321, 0.1802],
        [0.2053, 0.1287]])
c: tensor([[0.0161],
        [0.0219],
        [0.0220]], grad_fn=<AddBackward0>)
```

File: deepctr_torch/layers/core.py [hunk 1]
```diff
@@ -49,6 +51,10 @@ def __init__(self, inputs_dim, hidden_units, activation=F.relu, l2_reg=0, dropou
         if self.use_bn:
             self.bn = nn.ModuleList(
                 [nn.BatchNorm1d(hidden_units[i + 1]) for i in range(len(hidden_units) - 1)])
+        
+        self.activation_layers = nn.ModuleList(
+            [activation_layer(self.activation, hidden_units[i + 1]) for i in range(len(hidden_units) - 1)])
+
```
Comment (JyiHUO) [inline deepctr_torch/layers/core.py]: Making the code less robust

File: deepctr_torch/layers/core.py [hunk 2]
```diff
@@ -65,7 +71,7 @@ def forward(self, inputs):
             if self.use_bn:
                 fc = self.bn[i](fc)
 
-            fc = self.activation(fc)
+            fc = self.activation_layers[i](fc)
```
Comment (JyiHUO) [inline deepctr_torch/layers/core.py]: Making the code less robust. You should make your code more robust by using `self.activation`

File: deepctr_torch/layers/core.py [hunk 3]
```diff
@@ -4,6 +4,8 @@
 import torch.nn as nn
 import torch.nn.functional as F
 
+from .activation import activation_layer
+
```
Comment (JyiHUO) [inline deepctr_torch/layers/core.py]: Do not approve

File: deepctr_torch/layers/activation.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import torch.nn.functional as F
+
+
+device = 'cuda:0' if torch.cuda.is_available() else 'cpu'
+
+class Dice(nn.Module):
+    """
+    
+    References
+        - [Zhou G, Zhu X, Song C, et al. Deep interest network for click-through rate prediction[C]//Proceedings of the 24th ACM SIGKDD International Conference on Knowledge Discovery & Data Mining. ACM, 2018: 1059-1068.](https://arxiv.org/pdf/1706.06978.pdf)
```
Comment (JyiHUO) [inline deepctr_torch/layers/activation.py]: Do not add another file in `layer`

File: deepctr_torch/models/basemodel.py [hunk 5]
```diff
@@ -230,8 +231,13 @@ def fit(self, x=None,
...
                                     train_result[name] = []
-                                train_result[name].append(metric_fun(
-                                    y.cpu().data.numpy(), y_pred.cpu().data.numpy()))
+
+                                if use_double:
+                                    train_result[name].append(metric_fun(
+                                        y.cpu().data.numpy(), y_pred.cpu().data.numpy().astype("float64")))
+                                else:
+                                    train_result[name].append(metric_fun(
+                                        y.cpu().data.numpy(), y_pred.cpu().data.numpy()))
```
Comment (JyiHUO) [inline deepctr_torch/models/basemodel.py]: Good! Solved the issue

File: deepctr_torch/models/basemodel.py [hunk 6]
```diff
@@ -424,12 +434,24 @@ def _get_loss_func(self, loss):
...
+    def _log_loss(self, y_true, y_pred, eps=1e-7, normalize=True, sample_weight=None, labels=None):
+        # change eps to improve calculation accuracy
+        return log_loss(y_true,
+                        y_pred,
+                        eps,
+                        normalize,
+                        sample_weight,
+                        labels)
+
+    def _get_metrics(self, metrics, set_eps=False):
```
Comment (JyiHUO) [inline deepctr_torch/models/basemodel.py]: Excellent! Solved the issue [55](https://github.com/shenweichen/DeepCTR-Torch/issues/55)

File: deepctr_torch/layers/activation.py (hunk 4 above)
Comment (ZhangYuef) [reply deepctr_torch/layers/activation.py]: > Do not add another file in `layer`

I put this file here after referred to tensorflow version's file structure. So, where should I add the new activation layer?

File: deepctr_torch/layers/core.py (hunk 1 above)
Comment (ZhangYuef) [reply deepctr_torch/layers/core.py]: Yes~ I am going to complete the code here to support all kinds of built-in activation layers in `PyTorch`.

**Comment by ZhangYuef** [comment]
- Fix calculate accuracy problem.
- Going to fix multiple activation layer support.
